akka-projection-persistence-query/src/main/scala/akka/persistence/query/TypedEventEnvelope.scala

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.query
+
+object TypedEventEnvelope {
+  def apply[Event](eventEnvelope: EventEnvelope): TypedEventEnvelope[Event] = {
+    new TypedEventEnvelope[Event](eventEnvelope)
+  }
+}
+
+final class TypedEventEnvelope[Event] private[akka] (eventEnvelope: EventEnvelope) {
```

Comment:
Good point. I was trying to avoid all the hashCode / equals.

Reply:
Should we avoid case classes? `EventEnvelope` was one, but we step out when we added the timestamp.

Reply:
If we suspect that we will add more to it in the future it's much easier if it's not a case class. I think this is a good candidate for that, not-case class.

Reply:
Refactor it to a simple class.

I didn't implement Product5 and all the machinery to make it look like a case class,  but I added `apply`, `unapply` and `copy` methods because those are quite useful.